cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -400,6 +412,135 @@ async function activateOption(option, loadBalancerDNS, loadBalancerHostedZone) {
+        let promise = clients.acm().describeCertificate({
+            CertificateArn: arn
+        }).promise().then(( { Certificate} ) => {
+            cachedServiceCertificates.push(Certificate);
+        });
+        promises.push(promise);
+    }
+    await Promise.all(promises);
+    return cachedServiceCertificates;
+}
+
+/**
+ * Retrieve validation options that will be unused by the service after deleting the certificate.
+ * An option is deemed pending unused if it validates only the certificate(s) pending deletion.
+ *
+ * @param {Set<String>} aliases is the subject alternative names of the certificate(s) pending deletion.
+ * @returns {Set<Object>} The options that will be unused by the service.
+ */
+async function unusedOptionsByService(aliases) {
+    // NOTE: `Set` uses object reference to identify elements. That is, two options with the same value will not be
+    // identify as the same option. Therefore, we use map with object string as keys to identify object with actual
+    // values instead of object reference.
+    let optionsInUse = new Map();
+    let optionsPendingUnusedByService = new Map();
+
+    let certificates = await serviceCertificates();
+    for (const {DomainName: domainName, SubjectAlternativeNames: sans, DomainValidationOptions: validationOptions} of certificates) {
+        if (domainName === certificateDomain && setEqual(aliases.add(certificateDomain), new Set(sans))) {
+            // This is a certificate pending deletion. It is:
+            // 1. Copilot-tagged, 2. use the default certificate domain name, and 3. uses the SANs that need to be deleted.
+            for (let option of validationOptions) {
+                optionsPendingUnusedByService.set(JSON.stringify(option), option);
+            }
+            continue;
+        }
+        // Otherwise, this is a certificate that we do not want to touch; validation options of this certificate are deemed in use.
+        for (let option of validationOptions) {
+            optionsInUse.set(JSON.stringify(option), option);
+        }
+    }
+    for (const optionInUseKey of optionsInUse.keys()) {
+        optionsPendingUnusedByService.delete(optionInUseKey);
+    }
+    let options = new Set();
+    for (const optionPendingUnused of optionsPendingUnusedByService.values()) {
+        options.add(optionPendingUnused);
+    }
+    return options;
+}
+
+/**
+ * Retrieve validation options that will be unused by any service.
+ *
+ * @param {Set<String>} aliases
+ * @param {String} loadBalancerDNS
+ * @returns {Promise<Set<any>>}
+ */
+async function unusedValidationOptions(aliases, loadBalancerDNS) {
+    let optionsPendingUnused = await optionsUnusedByService(aliases);
+    let promises = [];
+    for (let option of optionsPendingUnused) {
+        let domainName = option["DomainName"];
+        let hostedZoneID, route53Client;
+        try {
+            ({hostedZoneID, route53Client} = await domainResources(domainName));
```

Comment:
```suggestion
            {hostedZoneID, route53Client} = await domainResources(domainName);
```

Reply:
The round parenthesis is sort of required if `hostedZoneID` and `route53Client` are first declared, and then assigned values with deconstruction. Otherwise it doesn't know how to parse it : (